File: 0084-largest-rectangle-in-histogram/0084-largest-rectangle-in-histogram.cpp
class Solution {
public:
    int largestRectangleArea(vector<int>& arr) {
        int maxi=0;
        stack<int>st;

        for(int i=0;i<arr.size();i++){
            while(!st.empty() && arr[st.top()]>arr[i]){
                int element=st.top();
                st.pop();
                int nse=i;
                int pse=st.empty()?-1:st.top();
                maxi=max(maxi,(nse-pse-1)*arr[element]);
            }
            st.push(i);
        }
        while(!st.empty()){
            int nse=arr.size();
            int element=st.top();
            st.pop();
            int pse=st.empty()?-1:st.top();
            maxi=max(maxi,(nse-pse-1)*arr[element]);
        }
        return maxi;

    }
};

File: 0084-largest-rectangle-in-histogram/0084-largest-rectangle-in-histogram.py
class Solution:
    def largestRectangleArea(self, heights: List[int]) -> int:
        stack=[]
        maxi=0
        n=len(heights)

        for i in range(n+1):
            while stack and (i==n or heights[stack[-1]]>=heights[i]):
                h=heights[stack.pop()]

                if stack:
                    width=i-stack[-1]-1
                else:
                    width=i
                
                maxi=max(maxi,h*width)
            stack.append(i)
        
        return maxi